File: 백준/Silver/9012. 괄호/괄호.py
import sys
input = sys.stdin.readline

T = int(input())


def check(value):
    stk = []
    for c in value:
        if c == "(":
            stk.append(c)
        elif c == ")":
            if stk:
                if stk[-1] == "(":
                    stk.pop()
                else:
                    return False
            else:
                return False
    if stk:
        return False
    else:
        return True


for _ in range(T):
    tmp = input()
    if check(tmp):
        print("YES")
    else:
        print("NO")

File: node.js/백준/Silver/9012. 괄호/괄호.js
const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const t = Number(input[0]);

let output = [];

for (let i = 1; i <= t; i++) {
    // 괄호 저장
    let stack = [];
    const str = input[i].trim();
    for (c of str) {
        if (c === "(") {
            stack.push(c);
            continue;
        }

        if (stack.length < 0) {
            stack.push(c);
            continue;
        }

        const top = stack[stack.length - 1];
        if (top === "(") {
            stack.pop();
        } else {
            stack.push(c);
        }
    }
    const tmp = stack.length === 0 ? "YES" : "NO";
    output.push(tmp);
}

console.log(output.join("\n"));
